### PR TITLE
core: add SymbolOpInterface and leverage in interpreter

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: xdsl/traits.py [hunk 1]
```diff
@@ -110,3 +112,27 @@ def verify(self, op: Operation) -> None:
                     # too; in which case it will check itself.
                     if not child_op.has_trait(IsolatedFromAbove):
                         regions += child_op.regions
+
+
+class SymbolOpInterface(OpTrait):
+    @staticmethod
+    def get_sym_name(op: Operation) -> StringAttr:
```
Comment (compor) [inline xdsl/traits.py:119]: nit: maybe rename to `get_sym_attr_name` to match the MLIR version of this?

File: docs/Toy/toy/__main__.py [hunk 2]
```diff
@@ -83,7 +83,7 @@ def main(path: Path, emit: str):
     if emit == "interpret":
         interpreter = Interpreter(module_op)
         interpreter.register_implementations(ToyFunctions())
-        interpreter.run_module()
+        interpreter.run("main")
```
Comment (PapyChacal) [inline docs/Toy/toy/__main__.py:86]: Lovely ☺️

**Review by PapyChacal — APPROVED** [review]

File: xdsl/traits.py [hunk 3]
```diff
@@ -110,3 +112,27 @@ def verify(self, op: Operation) -> None:
...
                         regions += child_op.regions
+
+
+class SymbolOpInterface(OpTrait):
+    @staticmethod
+    def get_sym_attr_name(op: Operation) -> StringAttr:
+        """
+        Returns the symbol of the operation
+        """
+        return cast(StringAttr, op.attributes["sym_name"])
```
Comment (AntonLydike) [inline xdsl/traits.py:123]: I'm not a huge fan of cast, I would prefer an `isa` guard instead, as cast won't fail at runtime.

`cast` is basically `type: ignore`

**Review by AntonLydike — APPROVED** [review]
Why are we using `StringAttr` here instead of `SymbolRefAttr`?

File: xdsl/traits.py [hunk 4]
```diff
@@ -110,3 +112,27 @@ def verify(self, op: Operation) -> None:
                     # too; in which case it will check itself.
                     if not child_op.has_trait(IsolatedFromAbove):
                         regions += child_op.regions
+
+
+class SymbolOpInterface(OpTrait):
```
Comment (math-fehr) [inline xdsl/traits.py:117]: Can you document this class directly?
And write the requirements that the operation must satisfy as well.

File: xdsl/traits.py [hunk 5]
```diff
@@ -110,3 +112,27 @@ def verify(self, op: Operation) -> None:
...
+        """
+        return cast(StringAttr, op.attributes["sym_name"])
+
+    def verify(self, op: Operation) -> None:
+        # import builtin here to avoid circular import
+        from xdsl.dialects.builtin import StringAttr
+
+        if "sym_name" not in op.attributes:
+            raise VerifyException(
+                f"Operation {op.name} must have sym_name to conform to {SymbolOpInterface.__name__}"
```
Comment (math-fehr) [inline xdsl/traits.py:131]: More than 88 characters here

Comment (math-fehr) [reply xdsl/traits.py:131]: Also, "must have sym_name attribute", to make that more explicit.

File: xdsl/traits.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/traits.py:117]: - [ ] document SymbolOpInterface

File: xdsl/traits.py (hunk 5 above)
Comment (superlopuh) [reply xdsl/traits.py:131]: I don't think it's 88 characters that we have in black. 
- [ ] look into splitting the line
- [ ] reword comment

File: xdsl/traits.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/traits.py:123]: That was my intended behaviour, the verification below already makes sure that it's a StrinAttr, I don't think we should be checking things in two places.

File: xdsl/interpreter.py [hunk 6]
```diff
@@ -208,6 +209,18 @@ class Interpreter:
...
+    @property
+    def symbol_table(self) -> dict[str, Operation]:
+        if self._symbol_table is None:
+            self._symbol_table = {}
+
+            for op in self.module.walk():
+                if op.has_trait(SymbolOpInterface):
+                    symbol = SymbolOpInterface.get_sym_attr_name(op)
+                    self._symbol_table[symbol.data] = op
+        return self._symbol_table
```
Comment (AntonLydike) [inline xdsl/interpreter.py:223]: How does this interact with the module being mutable?

File: xdsl/traits.py (hunk 3 above)
Comment (AntonLydike) [reply xdsl/traits.py:139]: I still think an additional assert is preferable to code that typechecks successfully but still fails (as there are plenty of cases when IR is unverified or in a broken state and still passed around).

I'd rather have the code blow up here, than somewhere down the line.

Comment (PapyChacal) [reply xdsl/traits.py:139]: I think the interpret should have a clear assumption that the IR is verified, and not check everything again, too. That's the point of even having verifiers and verification infrastructure, I believe. All passes also work on the contract that they take verified IR.

Comment (PapyChacal) [reply xdsl/traits.py:139]: Off course this should be reflected in the `xdsl-run` tool, which should make sure to verify the Module before interpreting anything.

File: xdsl/interpreter.py (hunk 6 above)
Comment (PapyChacal) [reply xdsl/interpreter.py:223]: Only the symbol table is?
Ideally this is something shared in a SymboleTable Interface, that many operations should have.
It is an important structural dimension of MLIR we currently completely miss.
See https://mlir.llvm.org/docs/SymbolsAndSymbolTables/

Comment (PapyChacal) [reply xdsl/interpreter.py:223]: Maybe I missunderstood you question though :thinking:

Comment (AntonLydike) [reply xdsl/interpreter.py:223]: What if I query the symbol table in a rewrite, then add a symbol, then get the table again. AFAIU I will get a stale table, right?

Comment (AntonLydike) [reply xdsl/interpreter.py:223]: Oooooh, this is on the interpreter! Sorry, for some reason I thought it was on the Module. Nevermind!

File: xdsl/traits.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/traits.py:139]: Oh wow totally missed that, my bad!

File: xdsl/interpreter.py (hunk 6 above)
Comment (PapyChacal) [reply xdsl/interpreter.py:223]: Ahah crossed mistake with mine on the other comment 😂 Seems like all is alright then 🙂

**Pull request merged**